[SYCL][E2E] Add functionality to split build and run of e2e tests

ayylol opened the pull request:
Adds functionality to split the compilation and execution of e2e tests across separate machines.

This functionality is enabled via the `test-mode` lit parameter. By default this is set to `full` and tests are built and ran on the same system, just like before. setting `test-mode` to either `run-only` or `build-only` enables the test splitting.

Two new lit features have been added: `run-mode` which is enabled when either in `run-only` or `full` testing mode, and `build-and-run-mode` which is only enabled when in `full` testing mode.

When in `build-only` mode all tests that can be built on the system will be built. When running a lit test in this mode two key things change:
- All `RUN:` lines will be executed unless they contain `%{run}`, `%{run-unfiltered-devices}` or `%if run-mode`.
- Unsupported and requires statements are ignored. Currently the only way to mark tests as unsupported in build only mode is to include `build-and-run-mode` in a requires statement, or use `UNSUPPORTED: true`.

When in `run-only` mode tests are not built, they are only executed. Deciding whether a test is supported in this mode works the same as in `full` mode. To only execute the tests we ignore all lit `RUN:` lines unless they contain `%{run}`, `%{run-unfiltered-devices}` or `%if run-mode`. Since we do not build the tests in this mode, for any test to pass we must have ran the tests in either `build-only` or `full` modes.

### Some notes/current limitations:
- Currently only the spir64 triple is supported for the build only mode.
- If a test is able to build, but fails during running we need to mark it as `XFAILS: run-mode` so that it does not `XPASS` when in build-only mode.
- The build-only mode can be ran manually on CI with the `SYCL E2E` action, selecting the "Linux, build" runner, and adding `--param test-mode=build-only` to `LIT_OPTS`. This pr does not have the needed CI changes to be able to use the run only mode.

ayylol commented:
@uditagarwal97 @sarnex @aelovikov-intel

Closed #15728 and reopened in this pr so that it is not a sycl-devops-pr and runs unnecessary checks.

Going to leave this marked as draft until you guys review and are ok with the changes in format.py and lit.cfg.py, then mark as open so that it pings the tests owners after

https://github.com/intel/llvm/actions/runs/11725963631 is the latest dry-run of build only

ayylol commented:
also going to clean up the git log and force push

sarnex commented:
@ayylol Sorry just to double check, is this ready for review from us CI people or is it still in progress? Thx.

ayylol commented:
> @ayylol Sorry just to double check, is this ready for review from us CI people or is it still in progress? Thx.

@sarnex Yep, its ready for review. Im just leaving it as draft in case any of the changes to the underlying logic from addressing reviews leads to more/different test changes.

sarnex reviewed (commented):
some basic comments

ayylol commented on sycl/test-e2e/AmdNvidiaJIT/kernel_and_bundle.cpp:
```diff
@@ -1,5 +1,6 @@
 // UNSUPPORTED: windows
 // REQUIRES: cuda || hip
+// REQUIRES: unsplit-mode
```
Will address this comment last to leave open to bike-shedding. Would something like `unsplit-test-mode` work? or (`unsplit-testing`), should the other features (`build-mode`, `run-mode`) follow suit? or are they clear enough.

I don't think i'd want to put the word "split" in this particular feature, since it is exactly what it isn't.

sarnex replied:
yeah of course i dont want to start bikeshedding, `REQUIRES: unsplit-test-mode` is fine with me

uditagarwal97 replied:
Won't it be clearer to have something like `only-build-test`, `only-run-test`, `build-and-run-test` instead of `build-mode`, `run-mode`, and `unsplit-test-mode`?

ayylol replied:
> Won't it be clearer to have something like `only-build-test`, `only-run-test`, `build-and-run-test` instead of `build-mode`, `run-mode`, and `unsplit-test-mode`?

The `run-mode` feature is useful because it allows us to have common behaviour between the run-only and unsplit modes, if we didnt have it we would have to mark tests that xfail at run-time as both XFAIL on `only-run-test` and `build-and-run-test`.
 
The `build-mode` feature I could not think of a valid use for, since if we have an xfail at build time we would always also want to mark it as an xfail at run-time since we wouldn't have an executable, also a test in the e2e folder that uses `REQUIRES: build-mode` seems a bit suspicious.

I added something like `build-only`/`run-only`/`unsplit` as a variable within the python scripts, but not as a feature since their usefulness in that context is unclear to me. Rather I'm using them to clean up if statements in the python scripts where we want to check what split-mode we are in.

ayylol commented on sycl/test-e2e/DeviceArchitecture/device_architecture_comparison_on_device_aot.cpp:
```diff
@@ -1,6 +1,6 @@
 // REQUIRES: arch-intel_gpu_pvc, ocloc
 
-// XFAIL: arch-intel_gpu_pvc
+// XFAIL: arch-intel_gpu_pvc, run-mode
```
should be `&&`, will change

ayylol replied:
actually just removed the arch-intel_gpu_pvc, since the test also requires it its basically saying that it always xfails when running

ayylol commented on sycl/test-e2e/InvokeSimd/Spec/ImplicitSubgroup/tuple.cpp:
```diff
@@ -1,5 +1,5 @@
 // TODO: enable when Jira ticket resolved
-// XFAIL: gpu
+// XFAIL: *
```
these tests fail when compiling, and since device features, like `gpu` are not available during building it is registered as a failure instead of xfail when in split build only mode when running the tests without this change.

sarnex replied:
ah ok the tests compfail, thanks

ayylol commented on sycl/test-e2e/Matrix/SG32/joint_matrix_colA_rowB_colC.cpp:
```diff
@@ -11,7 +11,7 @@
 // RUN: %{run} %t.out
 
-// XFAIL:*
+// XFAIL: run-mode
```
These changes are to differentiate between tests that fail when compiling vs tests that fail when running. These tests originally are able to build so marking as `XFAIL: run-mode` makes sure these tests arent reported as `XPASS` when building only.

aelovikov-intel replied:
I want (very strongly) to start with explicit tests markup first before we introduce any heuristics in the lit logic, especially if that's really heuristics and not some bullet-proof correct logic.

sarnex replied:
i guess im just worried this feature will slow down/confuse developers writing tests, im willing to proceed with this as-is and see what happens, but we should be ready to revert/fix the problem fast if some issue comes up

ayylol replied:
> i guess im just worried this feature will slow down/confuse developers writing tests, im willing to proceed with this as-is and see what happens, but we should be ready to revert/fix the problem fast if some issue comes up

For now since there is no automatic running on CI enabled, we'll just have to monitor the failures manually. 

However once this is integrated into the post/pre commit I think it should be pretty straightforward to figure out the appropriate action given the ci's behaviour: 
- If we see that a test passes the build stage but fails the run stage we have a runfail, and if this is expected we would mark it as XFAIL/UNSUPPORTED on `run-mode`. 
- If it fails on both the build and run stage then we have a compfail and we would mark as XFAIL/UNSUPPORTED * (Or we could mark as XFAIL for a feature that is non-device-specific, like `linux`).

ayylol commented on sycl/test-e2e/format.py:
```diff
@@ -154,18 +154,26 @@ def execute(self, test, litConfig):
-            (backend, _) = sycl_device.split(":")
-            triples.add(get_triple(test, backend))
+        if "run-mode" not in test.config.available_features:
+            if "unsplit-mode" in test.requires or "TEMPORARY_DISABLED" in test.requires:
```
This bit is for build-only mode, so that only the tests that are marked as `REQUIRES: unsplit-mode` or `REQUIRES: TEMPORARY_DISABLED` are reported as unsupported.

I guess it might be useful if I added a feature for run-only and build-only to avoid confusion in if statements like this

sarnex replied:
ah sorry, i missed that we were checking the test requires

ayylol replied:
I added the variable `test.config.split_mode` to hopefully make these types of if statements less confusing to read. (replaced `"run-mode" not in available_features` to `split_mode == "build-only"` )

aelovikov-intel commented:
TBH, I don't like "*split*"... Maybe it should be `test_mode := build-only|run-only|full` or similar? @uditagarwal97 @sarnex WDYT?

ayylol commented:
> TBH, I don't like "_split_"... Maybe it should be `test_mode := build-only|run-only|full` or similar? @uditagarwal97 @sarnex WDYT?

Ok, changed `split-mode` to `test-mode` and `unsplit-test-mode` to `run-and-build-mode`. Will update the commit message to reflect these changes.

ayylol commented:
Result of the latest build-only run: https://github.com/intel/llvm/actions/runs/11860224770
A single test failed (`WorkGroupMemory/swap_test.cpp`) However this is due the SYCL E2E action using the nightly build and the test having changes that were done [this morning](https://github.com/intel/llvm/commit/d76b5529336d594bccbc9c44436aef038f4e8301), so if I run it after today it should pass.

@uditagarwal97 @aelovikov-intel @sarnex do the changes to format.py and lit.cfg.py look good? If so i'll open for review

aelovikov-intel commented on sycl/test-e2e/lit.cfg.py:
```diff
@@ -37,6 +37,20 @@
 config.required_features = []
 config.unsupported_features = []
 
+# split-mode: Set if tests should run normally or with split build/run
```
nit: `split-mode` isn't used anymore, comment needs to be updated.

uditagarwal97 commented on sycl/test-e2e/lit.cfg.py:
```diff
@@ -37,6 +37,20 @@
+        config.test_mode = "build-only"
+        config.sycl_devices = []
+        arch_flag = ""
+    case _:
```
`test-mode` can only accept `run-only`, `build-only`, and `full`, right? If so, shouldn't we do something like the following:
```
case "run-only":
   ...
case "build-only":
  ...
case "full":
  ...
case _:
  print("Invalid argument to test-mode")
  exit(1)
```

ayylol replied:
ok, done in latest commit

uditagarwal97 reviewed (commented):
`lit.cfg.py` and `format.py` changes LGTM too. Just one minor comment.

ayylol commented:
Closed the following PRs to include their changes in this PR: #15960, #15950, #15891, #15787, #15789

Doing this all in one pr so that all the changes needed to cleanly run the build-only mode on ci are included in one single commit.

YuriPlyakhin reviewed (approved):
Joint Matrix changes LGTM

againull commented:
@intel/bindless-images-reviewers @intel/syclcompat-lib-reviewers @intel/unified-runtime-reviewers   Friendly ping.

ayylol commented on sycl/test-e2e/DeviceArchitecture/device_architecture_comparison_on_device_aot.cpp:
```diff
@@ -1,6 +1,6 @@
 // REQUIRES: arch-intel_gpu_pvc, ocloc
 
-// XFAIL: arch-intel_gpu_pvc
+// XFAIL: run-mode
```
Yep, because the test requires `arch-intel_gpu_pvc` this essentially ended up being `XFAIL: *` as far as I am aware.
